src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTree`1.cs

```diff
@@ -35,24 +36,71 @@ internal partial class IntervalTree<T> : IEnumerable<T>
         where TIntrospector : struct, IIntervalIntrospector<T>
     {
         var result = new IntervalTree<T>();
 
-        AddAll(in introspector, values1);
-        AddAll(in introspector, values2);
+        if (values != null)
+        {
+            foreach (var value in values)
+                result.root = Insert(result.root, new Node(value), in introspector);
+        }
 
         return result;
+    }
 
-        void AddAll(in TIntrospector introspector, IEnumerable<T>? values)
+    /// <summary>
+    /// Creates an interval tree from a sorted list of values.  This is more efficient than creating from an unsorted
+    /// list as building doesn't need to figure out where the nodes need to go n-log(n) and doesn't have to rebalance
+    /// anything (again, another n-log(n) operation).  Rebalancing is particularly expensive as it involves tons of
+    /// pointer chasing operations, which is both slow, and which impacts the GC which has to track all those writes.
+    /// </summary>
+    /// <remarks>
+    /// The values must be sorted such that given any two elements 'a' and 'b' in the list, if 'a' comes before 'b' in
+    /// the list, then it's "start position" (as determined by the introspector) must be less than or equal to 'b's
+    /// start position.  This is a requirement for the algorithm to work correctly.
+    /// </remarks>
+    public static IntervalTree<T> CreateFromSorted<TIntrospector>(in TIntrospector introspector, SegmentedList<T> values)
+        where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+#if DEBUG
+        if (values.Count >= 2)
         {
-            if (values != null)
+            var current = values[0];
+            for (var i = 1; i < values.Count; i++)
             {
-                foreach (var value in values)
-                    result.root = Insert(result.root, new Node(value), in introspector);
+                var next = values[i];
+                Debug.Assert(introspector.GetSpan(current).Start <= introspector.GetSpan(next).Start);
+                current = next;
             }
         }
+#endif
+
+        if (values.Count == 0)
+            return Empty;
+
+        return new IntervalTree<T>
+        {
+            root = CreateFromSortedWorker(values, 0, values.Count, in introspector),
+        };
+    }
+
+    private static Node? CreateFromSortedWorker<TIntrospector>(
+        SegmentedList<T> values, int startInclusive, int endExclusive, in TIntrospector introspector) where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+        var length = endExclusive - startInclusive;
+        if (length <= 0)
+            return null;
+
+        var mid = startInclusive + (length >> 1);
+        var node = new Node(values[mid]);
+        node.SetLeftRight(
+            CreateFromSortedWorker(values, startInclusive, mid, in introspector),
+            CreateFromSortedWorker(values, mid + 1, endExclusive, in introspector),
+            in introspector);
+
+        return node;
```

Comment:
for anyone reviewing this.  To convince yourself of correctness, simply see how 'Insert' works:

```c#
    private static Node Insert<TIntrospector>(Node? root, Node newNode, int newNodeStart, in TIntrospector introspector)
        where TIntrospector : struct, IIntervalIntrospector<T>
    {
        if (root == null)
        {
            return newNode;
        }

        Node? newLeft, newRight;

        if (newNodeStart < introspector.GetSpan(root.Value).Start)
        {
            newLeft = Insert(root.Left, newNode, newNodeStart, in introspector);
            newRight = root.Right;
        }
        else
        {
            newLeft = root.Left;
            newRight = Insert(root.Right, newNode, newNodeStart, in introspector);
        }

        root.SetLeftRight(newLeft, newRight, in introspector);
        var newRoot = root;

        return Balance(newRoot, in introspector);
    }
```

The logic there is trivial.  We simply examine the start of the element we're adding and the start of the node we have to determine down which path to place the new node.    So that same 'sorting' is accomplished by an up front real sort of the items.

Then, once we've placed things down the correct path, 'set' the left/right elements (which picks the right height value for the node we're creating, as well as the right 'max end node').  That same 'setting' happens in the new code.

What chnages those is that we need no balancing.  because we're always building from another sorted sub-node, so the tree is naturally balanced enough to efficiently search without any costly steps while building it.

Reply:
I need to think this through a bit, but I'll ask before I reach understanding of this code: Why couldn't this just store an array and not need the actual tree structure?

Reply:
We just had that conversation in person :-)

Reply:
Sometimes I just nod and act like I'm listening to you :)